sympy/matrices/common.py

```diff
@@ -2989,6 +2991,46 @@ def __iter__(self): # supports numpy.matrix and numpy.array
         return iter(sympify(mat[r, c]) for r in range(self.rows) for c in range(cols))
 
 
+class MatrixKind(Kind):
+    """
+    Kind for all matrices in SymPy.
+
+    Basic class for this kind is ``MatrixBase`` and ``MatrixExpr``,
+    but any expression representing the matrix can have this.
+
+    Parameters
+    ==========
+
+    element_kind : Kind
+        Kind of the element. Default is ``NumberKind``, which means that
+        the matrix contains only numbers.
+
+    Examples
+    ========
+
+    >>> from sympy import MatrixSymbol, Integral
+    >>> from sympy.abc import x
+    >>> A = MatrixSymbol('A', 2,2)
+    >>> A.kind
+    MatrixKind(NumberKind)
+    >>> Integral(A,x).kind
+    MatrixKind(NumberKind)
+
+    See Also
+    ========
+
+    sympy.tensor.ArrayKind : Kind for N-dimensional arrays.
```

Comment:
Could be a see also here for `shape` and also an example showing how to use it.

It's probably worth mentioning that `shape(obj)` is expected to work whenever the `kind` is `MatrixKind`. It is the key property that is implemented for this `kind`.

Reply:
We may have more functions like `shape` (for example, we may need `rank` function) in the future, and it will be unnecessary to mention every functions in each `Kind`'s docstrings. I think mentioning `shape()` in tutorial doc will suffice.

Reply:
Somewhere there should be a list of the basic things that you can do with objects of a given `Kind`. In general we should have a lot more "See also" cross-referencing in the docs because it is very useful for someone reading the docs who does not already know the codebase.